### PR TITLE
client,daemon,cmd/snap: change POST /2.0/interfaces to work with lists

### PR DESCRIPTION
**Pull request by zyga**
This patch changes the REST API for interacting with interfaces to
operate on lists of plugs and slots. In practice nothing changes but
this patch opens the avenue for connecting or disconnecting multiple
plugs and sockets in one call.

The connect/disconnect/{add,remove}-{plug,slot} actions are constrained
to one plug/slot. This restriction can be lifted later by another patch,
without any changes to the wire format.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: daemon/api.go [hunk 1]
```diff
@@ -949,19 +949,28 @@ func changeInterfaces(c *Command, r *http.Request) Response {
 	}
 	switch a.Action {
 	case "connect":
-		err := c.d.interfaces.Connect(a.Plug.Snap, a.Plug.Name, a.Slot.Snap, a.Slot.Name)
+		if len(a.Plugs) != 1 || len(a.Slots) != 1 {
```
Comment (chipaca) [inline daemon/api.go]: I think these checks are all wrong :-)

That is: you correctly check for `!= 1`, but you error return assumes `>1`.

Maybe you can check for `>1` outside of the case, return the "not implemented" 400 there, and then check for the `<1` in the cases where you need to check, and return the appropriate "missing" 400 in each one?

**Comment by chipaca** [comment]
p.s., “client,daemon,cmd/snap” would be a lot more helpful for @elopio et al than “many”, aiui.

File: docs/rest.md [hunk 2]
```diff
@@ -670,7 +670,11 @@ Sample input:
...
 {
     “action”: “connect”,
-    “plug”: {“snap”: “canonical-pi2”,   “plug”: “pin-13”},
-    “slot”: {“snap”: “keyboard-lights”, “slot”: “capslock-led”}
+    “plugs”: {
+        {“snap”: “canonical-pi2”,   “plug”: “pin-13”},
+    },
+    “slots”: {
+        {“snap”: “keyboard-lights”, “slot”: “capslock-led”},
+    }
```
Comment (chipaca) [inline docs/rest.md]: `"plus": [...]`. That is: square brackets for lists.

Comment (chipaca) [inline docs/rest.md]: also, I didn't notice before, but typographic quotes have snuck in to the json examples. Please use straight quotes in code.

**Comment by chipaca** [comment]
:+1:

File: client/interfaces.go [hunk 3]
```diff
@@ -95,13 +95,17 @@ func (client *Client) performInterfaceAction(sa *InterfaceAction) error {
...
 		},
-		Slot: &Slot{
-			Snap: slotSnapName,
-			Name: slotName,
+		Slots: []Slot{
+			{
+				Snap: slotSnapName,
+				Name: slotName,
+			},
 		},
```
Comment (niemeyer) [inline client/interfaces.go]: The 11 lines above are actually just 2 lines (!!!).

File: client/interfaces.go [hunk 4]
```diff
@@ -110,13 +114,17 @@ func (client *Client) Connect(plugSnapName, plugName, slotSnapName, slotName str
...
```
Comment (niemeyer) [inline client/interfaces.go]: The 11 lines above are actually just 2 lines (!!!).

File: client/interfaces.go [hunk 5]
```diff
@@ -133,9 +141,11 @@ func (client *Client) AddPlug(plug *Plug) error {
...
 	return client.performInterfaceAction(&InterfaceAction{
 		Action: "remove-plug",
-		Plug: &Plug{
-			Snap: snapName,
-			Name: plugName,
+		Plugs: []Plug{
+			{
+				Snap: snapName,
+				Name: plugName,
+			},
```
Comment (niemeyer) [inline client/interfaces.go]: These 6 lines are actually a single one (!!!).

File: client/interfaces.go [hunk 6]
```diff
@@ -152,9 +162,11 @@ func (client *Client) AddSlot(slot *Slot) error {
...
 	return client.performInterfaceAction(&InterfaceAction{
 		Action: "remove-slot",
-		Slot: &Slot{
-			Snap: snapName,
-			Name: slotName,
+		Slots: []Slot{
+			{
+				Snap: snapName,
+				Name: slotName,
+			},
```
Comment (niemeyer) [inline client/interfaces.go]: These 6 lines are actually a single one (!!!).

File: daemon/api_test.go [hunk 7]
```diff
@@ -1400,13 +1408,17 @@ func (s *apiSuite) TestConnectPlugFailureNoSuchPlug(c *check.C) {
...
+			},
 		},
-		Slot: interfaces.Slot{
-			Snap: "consumer",
-			Name: "slot",
+		Slots: []interfaces.Slot{
+			{
+				Snap: "consumer",
+				Name: "slot",
+			},
```
Comment (niemeyer) [inline daemon/api_test.go]: Just two lines please.

File: daemon/api_test.go [hunk 8]
```diff
@@ -1438,13 +1450,17 @@ func (s *apiSuite) TestConnectPlugFailureNoSuchSlot(c *check.C) {
...
```
Comment (niemeyer) [inline daemon/api_test.go]: Just two lines please.

File: daemon/api_test.go [hunk 9]
```diff
@@ -1478,13 +1494,17 @@ func (s *apiSuite) TestDisconnectPlugSuccess(c *check.C) {
...
 	action := &interfaceAction{
 		Action: "disconnect",
-		Plug: interfaces.Plug{
-			Snap: "producer",
-			Name: "plug",
+		Plugs: []interfaces.Plug{
+			{
+				Snap: "producer",
+				Name: "plug",
+			},
```
Comment (niemeyer) [inline daemon/api_test.go]: Just two lines please.

File: daemon/api_test.go [hunk 10]
```diff
@@ -1478,13 +1494,17 @@ func (s *apiSuite) TestDisconnectPlugSuccess(c *check.C) {
...
 		},
-		Slot: interfaces.Slot{
-			Snap: "consumer",
-			Name: "slot",
+		Slots: []interfaces.Slot{
+			{
+				Snap: "consumer",
+				Name: "slot",
+			},
```
Comment (niemeyer) [inline daemon/api_test.go]: Just two lines please.

File: daemon/api_test.go [hunk 11]
```diff
@@ -1514,13 +1534,17 @@ func (s *apiSuite) TestDisconnectPlugFailureNoSuchPlug(c *check.C) {
...
```
Comment (niemeyer) [inline daemon/api_test.go]: Just two lines please.

File: daemon/api_test.go [hunk 12]
```diff
@@ -1552,13 +1576,17 @@ func (s *apiSuite) TestDisconnectPlugFailureNoSuchSlot(c *check.C) {
...
```
Comment (niemeyer) [inline daemon/api_test.go]: Just two lines please.

File: daemon/api_test.go [hunk 13]
```diff
@@ -1591,13 +1619,17 @@ func (s *apiSuite) TestDisconnectPlugFailureNotConnected(c *check.C) {
...
+				Name: "plug",
+			},
 		},
-		Slot: interfaces.Slot{
-			Snap: "consumer",
-			Name: "slot",
+		Slots: []interfaces.Slot{
+			{
+				Snap: "consumer",
+				Name: "slot",
```
Comment (niemeyer) [inline daemon/api_test.go]: Just two lines please.

File: docs/rest.md [hunk 14]
```diff
@@ -670,7 +670,11 @@ Sample input:
 ```javascript
 {
     “action”: “connect”,
-    “plug”: {“snap”: “canonical-pi2”,   “plug”: “pin-13”},
-    “slot”: {“snap”: “keyboard-lights”, “slot”: “capslock-led”}
+    “plugs”: {
+        {“snap”: “canonical-pi2”,   “plug”: “pin-13”},
+    },
+    “slots”: {
+        {“snap”: “keyboard-lights”, “slot”: “capslock-led”},
```
Comment (niemeyer) [inline docs/rest.md]: Just two lines please.

File: daemon/api_test.go [hunk 15]
```diff
@@ -2036,9 +2090,11 @@ func (s *apiSuite) TestRemoveSlotFailure(c *check.C) {
...
 	action := &interfaceAction{
 		Action: "remove-slot",
-		Slot: interfaces.Slot{
-			Snap: "consumer",
-			Name: "slot",
+		Slots: []interfaces.Slot{
+			{
+				Snap: "consumer",
+				Name: "slot",
+			},
```
Comment (niemeyer) [inline daemon/api_test.go]: Single line.

File: daemon/api_test.go [hunk 16]
```diff
@@ -2001,9 +2053,11 @@ func (s *apiSuite) TestRemoveSlotDisabled(c *check.C) {
 	d.enableInternalInterfaceActions = false
 	action := &interfaceAction{
 		Action: "remove-slot",
-		Slot: interfaces.Slot{
-			Snap: "consumer",
-			Name: "slot",
+		Slots: []interfaces.Slot{
+			{
+				Snap: "consumer",
+				Name: "slot",
```
Comment (niemeyer) [inline daemon/api_test.go]: Single line.

File: daemon/api_test.go [hunk 17]
```diff
@@ -1969,9 +2019,11 @@ func (s *apiSuite) TestRemoveSlotSuccess(c *check.C) {
 	d.interfaces.AddSlot(&interfaces.Slot{Snap: "consumer", Name: "slot", Interface: "interface"})
 	action := &interfaceAction{
 		Action: "remove-slot",
-		Slot: interfaces.Slot{
-			Snap: "consumer",
-			Name: "slot",
+		Slots: []interfaces.Slot{
+			{
+				Snap: "consumer",
+				Name: "slot",
```
Comment (niemeyer) [inline daemon/api_test.go]: Single line.

File: daemon/api_test.go [hunk 18]
```diff
@@ -1930,15 +1978,17 @@ func (s *apiSuite) TestAddSlotFailure(c *check.C) {
...
-			Attrs: map[string]interface{}{
-				"key": "value",
+		Slots: []interfaces.Slot{
+			{
+				Snap:      "snap",
+				Name:      "slot",
+				Label:     "label",
+				Interface: "interface",
+				Attrs: map[string]interface{}{
+					"key": "value",
```
Comment (niemeyer) [inline daemon/api_test.go]: Single line.

File: daemon/api_test.go [hunk 19]
```diff
@@ -1887,15 +1933,17 @@ func (s *apiSuite) TestAddSlotDisabled(c *check.C) {
...
-			Attrs: map[string]interface{}{
-				"key": "value",
+		Slots: []interfaces.Slot{
+			{
+				Snap:      "consumer",
+				Name:      "slot",
+				Label:     "label",
+				Interface: "interface",
+				Attrs: map[string]interface{}{
+					"key": "value",
```
Comment (niemeyer) [inline daemon/api_test.go]: Single line.

File: daemon/api_test.go [hunk 20]
```diff
@@ -1848,15 +1892,17 @@ func (s *apiSuite) TestAddSlotSuccess(c *check.C) {
...
-			Attrs: map[string]interface{}{
-				"key": "value",
+		Slots: []interfaces.Slot{
+			{
+				Snap:      "snap",
+				Name:      "slot",
+				Label:     "label",
+				Interface: "interface",
+				Attrs: map[string]interface{}{
+					"key": "value",
```
Comment (niemeyer) [inline daemon/api_test.go]: Single line.

File: daemon/api_test.go [hunk 21]
```diff
@@ -1816,9 +1858,11 @@ func (s *apiSuite) TestRemovePlugFailure(c *check.C) {
...
 	action := &interfaceAction{
 		Action: "remove-plug",
-		Plug: interfaces.Plug{
-			Snap: "producer",
-			Name: "plug",
+		Plugs: []interfaces.Plug{
+			{
+				Snap: "producer",
+				Name: "plug",
+			},
```
Comment (niemeyer) [inline daemon/api_test.go]: Single line.

File: daemon/api_test.go [hunk 22]
```diff
@@ -1781,9 +1821,11 @@ func (s *apiSuite) TestRemovePlugDisabled(c *check.C) {
...
```
Comment (niemeyer) [inline daemon/api_test.go]: Single line.

File: daemon/api_test.go [hunk 23]
```diff
@@ -1749,9 +1787,11 @@ func (s *apiSuite) TestRemovePlugSuccess(c *check.C) {
...
```
Comment (niemeyer) [inline daemon/api_test.go]: Single line.

**Comment by niemeyer** [comment]
LGTM, but let's please fix the formatting. It got over the top recently. Splitting 2 simple lines in 11 is unreasonable.

File: docs/rest.md (hunk 2 above)
Comment (zyga) [inline docs/rest.md]: Corrected (and elsewhere in rest.md) this is a side effect of copy-pasting from google docs.

**Pull request merged**